### PR TITLE
add typescript definition

### PR DESCRIPTION
**Pull request by taoqf**

### CODE REVIEW

**Comment by pofider** [comment]
Thank you for your work. There are some things I would like to discuss about it.

Your single definition file now strictly define types, however jsreport is very much dynamic and final types are composition of core and installed extensions. The `jsreport-phantom-pdf` for example extends the template type with phantom property where the header or footer can be defined. It would be great if we can use [declaration files merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) or other typescript concept to reach the same in typescript types. Every extension would then provide its definition file which would be merged in.  The developer would then see proper types based on the extensions imported. We would also don't need to maintain one big files covering everything.

What you think about it? Would it be doable?

**Comment by bjrmatos** [comment]
hi! thanks for the PR. here is some things that come to my mind with this addition

- one thing to consider is the maintenance cost of this addition, i'm not a huge fan of having type definitions in a codebase which does not use types, feels weird and it will be more work for us to keep the definitions updated.

- i think that type definitions for repositories that does not use typescript at all should live in [definitelytyped](http://definitelytyped.org/) and should be maintained by typescript users which knows best how to express API in such language. we are not typescript users (in jsreport) so it feels weird to me if we are going to start maintaining types for something that we don't use (of course that can change in the future but just describing the present)

please don't see my comments as a rejection 😄, i'm just trying to describe the current situation.. 

thanks for your work, i'm sure that this will be a great start for having jsreport typed (whether if the types live in our own repository or in definitelytyped).

**Comment by pofider** [comment]
The discussion here kind of froze, but I would really like to continue with this because the start proposed in PR is good. The code will never be merged here and will in the end live in [definitelytyped](http://definitelytyped.org/), but that is not the reason to stop.

I think the main problem to resolve now is the definition files merging to allow extensions to participate on the final type as I described in [this comment](https://github.com/jsreport/jsreport-core/pull/26#issuecomment-353864470).

**Comment by taoqf** [comment]
Sorry for seeing your replies so late. I do not know much about jsreport, Is it something `jsreport` kind like `jquery`?  I've made a [pr](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/22913) in [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped).

**Comment by pofider** [comment]
The jquery plugins extends the core `$` type.
jsreport extensions extends the jsreport core types and enums.

Your PR defines for example the template type like this:
```js
template: {
		content: string;
		engine: 'jsrender' | 'handlebars' | 'ejs' | 'jade';
		helpers: string;
 		recipe: 'phantom-pdf' | 'electron-pdf' | 'text' | 'xlsx' | 'html-to-xlsx' | 'phantom-image' | 'html-to-text' | 'fop-pdf' | 'client-html' | 'wrapped-html' | 'wkhtmltopdf';
 	};
```
However this is not full type when using even single extension `jsreport-phantom-pdf` which extends the template with property `phantom: { header: string ..... }`. It would be great if every extension can add its own props to the core types using some kind of types merging. I believe it could be doable as this is common problem also with mentioned jquery.

Regarding the PR to DefinitelyTyped. I don't know what is the majority opinion here. I would rather close it and bake the final form here and as we are ready send the final form there. But maybe I'm wrong. In every case the current form needs some refinements and usage of merged types instead of naming everything in one place.

@taoqf Would you have time to play with the types merging? I've tried it quickly without success, but it should work somehow.

**Comment by taoqf** [comment]
Yes, I'd like to fix the types, infact 
```ts
recipe: 'phantom-pdf' | 'electron-pdf' | 'text' | 'xlsx' | 'html-to-xlsx' | 'phantom-image' | 'html-to-text' | 'fop-pdf' | 'client-html' | 'wrapped-html' | 'wkhtmltopdf' | string;
```
may be an easier way to define the types.

**Comment by taoqf** [comment]
@pofider , I can even give some test code here:
```ts
import * as inline from 'inline-source';
import * as JsReportCore from 'jsreport-core';

const jsreport = JsReportCore();

export default function html2pdf(html: string) {
	return new Promise<Buffer>((resolve, reject) => {
		inline(html, {
			compress: true,
			// rootpath: path.resolve('www'),
			// Skip all css types and png formats
			ignore: ['png']
		}, (err, content) => {
			jsreport.init().then(() => {
				return jsreport.render({
					template: {
						content,
						engine: 'jsrender',
						recipe: 'phantom-pdf'
					}
				}).then((resp) => {
					// fs.writeFileSync('./a.pdf', resp.content);
					resolve(resp.content);
				}, (e) => {
					reject(e);
				});
			}).catch((e) => {
				reject(e);
			});
		});
	});
}
```

**Comment by pofider** [comment]
The `... | string` makes it better. I see.

However now you want to do

```js
return jsreport.render({
   template: {
     content,
     engine: 'jsrender',
     recipe: 'phantom-pdf',
     phantom: { header: 'Hello world' }
  }
})
```
It would be great if we can merge types from phantom into the root template type.

**Comment by taoqf** [comment]
```ts
interface RenderOptions {
	template: {
		content: string;
		engine: 'jsrender' | 'handlebars' | 'ejs' | 'jade';
		helpers?: string | { [fun: string]: helpers };
		recipe: 'phantom-pdf' | 'electron-pdf' | 'text' | 'xlsx' | 'html-to-xlsx' | 'phantom-image' | 'html-to-text' | 'fop-pdf' | 'client-html' | 'wrapped-html' | 'wkhtmltopdf';
		phantom?: any;
	};
	data?: any;
}
```
or
```ts
interface RenderOptions {
	template: {
		content: string;
		engine: 'jsrender' | 'handlebars' | 'ejs' | 'jade';
		helpers?: string | { [fun: string]: helpers };
		recipe: 'phantom-pdf' | 'electron-pdf' | 'text' | 'xlsx' | 'html-to-xlsx' | 'phantom-image' | 'html-to-text' | 'fop-pdf' | 'client-html' | 'wrapped-html' | 'wkhtmltopdf';
		[key: string]: any;
	};
	data?: any;
}
```
I really do not know what you want to do.

**Comment by taoqf** [comment]
Are you guys would merge this pr? I have closed https://github.com/DefinitelyTyped/DefinitelyTyped/pull/22913 already. if not, I will reopen that pr.

**Comment by pofider** [comment]
I will try to explain on example again what I would like to reach. Please read my previous comments again to get the context.

To shortly repeat again. The problem here is that you create one giant file with definitions from all extensions (> 50). The downside of this is that everybody who creates and publish an extension needs to send a PR to single jsreport types. The second problem with one giant definition file is that user who only use 3 extensions gets types from all of 50 which is very misleading. It would be much better if the types are split into dedicated packages `@types/jsreport-core`, `@types/jsreport-scripts`, `@types/jsreport-images`....

Example
 
```js
// core.d.ts
interface Reporter {
    coreProp: string
}

declare function init (): Reporter;
export = init
```

```js
// extensionA.d.ts
interface Reporter {
    // adding additional property to the core Reporter interface
    extensionAProp: string
}

declare function extensionA(Reporter): Reporter;
export = extensionA
```

```js
// main.ts
import core = require("jsreport-core")
import extensionA = require("extensionA")

const reporter = core()
extensionA(reporter)
```
I would expect that the definition files `core.d.ts` and `extensionA.d.ts` are merged in this case and the final `Reporter` interface includes both props `coreProp` and `extensionAProp`.
The problem is that this doesn't work for me. The types merging works only if I put everything into the single file. Do you know if this could work somehow?
The documentation for [declaration-merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) doesn't say anything how this could work if multiple declaration files are used. Thank you.

**Comment by taoqf** [comment]
Should this work? @pofider

**Comment by pofider** [comment]
Wonderful. Excellent work! Thank you.  This works for me. This is what I wanted.

This is so powerful that the jsreport-core can even specify the enum type for engine as

```js
enum Engine {
    none = 1
}

interface Template {
    content?: string;
    engine?: Engine | string;       
}

interface RenderOptions {
    template: Template,
    data?: any;
}
```

and the `jsreport-jsrender` extension can even its own value to the enum as

```js
declare namespace JsReport {
	enum Engine {
		jsrender
	}
}
```

And the intellisense works nice afterwards

image

I was a bit afraid of this, but as this works everything else should as well.

To further proceed, we will need to cut out some of the types from the main jsreport core and move them to the respective extensions. I think I will send a minor PR to the [@types/jsreport-core](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/jsreport-core) and then we can create the types for the popular extensions.

**Comment by taoqf** [comment]
I would rather we could maintain type definitions in this project.

**Comment by pofider** [comment]
I've finished the changes in definition files and sent you [PR](https://github.com/taoqf/jsreport-core/pull/1) for review if you find time, please look at it. Thank you.

> I would rather we could maintain type definitions in this project.

This would be easier for us. However the common practice is different. The projects that doesn't use typescript mostly store types externally and it seems to me reasonable.

**Comment by taoqf** [comment]
I have tried my best to take a preview.

> This would be easier for us. However the common practice is different. The projects that doesn't use typescript mostly store types externally and it seems to me reasonable.

maybe these projects who doesn't care about typescript should put them into [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped). or we could contribute, if the authors would like to merge the prs

**Comment by taoqf** [comment]
pls see https://github.com/jsreport/jsreport-jsrender/pull/2 https://github.com/jsreport/jsreport-phantom-pdf/pull/25 https://github.com/jsreport/jsreport-html-to-xlsx/pull/14 and https://github.com/jsreport/jsreport-xlsx/pull/31 thank you. @pofider

File: types/jsreport-core/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+declare namespace JsReport {
+  type Helpers = string | { [fun: string]: (...args: any[]) => any };
+
+  const enum EngineType {
+    None = "none"
+  }
+
+  const enum RecipeType {
+    // Html = "html"
```
Comment (pofider) [inline types/jsreport-core/index.d.ts]: Why is this removed? There is html recipe right in the jsreport core [here](https://github.com/jsreport/jsreport-core/blob/master/lib/render/htmlRecipe.js).

File: types/jsreport-html-to-xlsx/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	const enum RecipeType {
+		Html2Xlsx = 'html-to-xlsx'
+	}
+
+	interface Html2Xlsx {
+		shortid: string;
+	}
+
+	interface Template {
+		xlsxTemplate: Partial<Html2Xlsx>;
```
Comment (pofider) [inline types/jsreport-html-to-xlsx/index.d.ts]: The `html-to-xlsx` recipe adds just recipe type.  It doesn't extend the template. See the [source](https://github.com/jsreport/jsreport-html-to-xlsx/blob/master/lib/htmlToXlsx.js). You should keep here only the RecipeType changes.

File: types/jsreport-phantom-pdf/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    height: string;
+    headerHeight: string;
+    footerHeight: string;
+    format: string;
+    orientation: 'portrait' | 'landscape';
+    blockJavaScript: 'true' | any;
+    resourceTimeout: number;
+    waitForJS: 'true' | 'false';
+    fitToPage: 'true' | 'false';
+    customPhantomJS: 'true' | 'false';
```
Comment (pofider) [inline types/jsreport-phantom-pdf/index.d.ts]: The boolean types should be defined as boolean. Or as 'true' | 'false' | boolean.
See the meta types declared in jsreport [here](https://github.com/jsreport/jsreport-phantom-pdf/blob/master/lib/phantom.js#L42).

File: types/jsreport-phantom-pdf/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  }
+}
+
+declare namespace JsReportPhantomPdf {
+  interface Options {
+    allowLocalFilesAccess: boolean;
+    appDirectory: string;
+    defaultPhantomjsVersion: string;
+    strategy: string;
+  }
```
Comment (pofider) [inline types/jsreport-phantom-pdf/index.d.ts]: Please include `timeout: number`. The `appDirectory` on the other hand should not be there. Could the strategy be also an enum? 
```js
const enum PhantomStrategy {
  dedicatedProcess = 'dedicated-process',
  phantomServer = 'phantom-server'
}
```

File: types/jsreport-core/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+// Definitions by: taoqf <https://github.com/taoqf>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.3
+
+/// <reference types="node" />
+
+declare namespace JsReport {
+  type Helpers = string | { [fun: string]: (...args: any[]) => any };
+
+  const enum EngineType {
```
Comment (pofider) [inline types/jsreport-core/index.d.ts]: Any reason why you postfix enums with 'Type'? I mean why you changed `Engine` to `EngineType` for example? I see in typescript [convention style guide](https://github.com/basarat/typescript-book/blob/master/docs/styleguide/styleguide.md#enum) that enums are without postfix. Like `Color` instead of `ColorType`.

Comment (taoqf) [reply types/jsreport-core/index.d.ts]: en.. i think may be we need a base interface called `Engine` and `Recipe`, sorry for that I could not name the interface line 59 and line 61. so, I add Type as the postfix.

File: types/jsreport-phantom-pdf/index.d.ts (hunk 3 above)
Comment (taoqf) [reply types/jsreport-phantom-pdf/index.d.ts]: I define this just because i saw [this](https://github.com/jsreport/jsreport-phantom-pdf/blob/master/lib/phantom.js#L152)

Comment (taoqf) [reply types/jsreport-phantom-pdf/index.d.ts]: I think if it is defined as `boolean`, may be cause an excaption?

File: types/jsreport-core/index.d.ts (hunk 5 above)
Comment (pofider) [reply types/jsreport-core/index.d.ts]: Ah ok. I didn't notice that. The `jsreport.use` function accepts function or object, so I'm not sure if we need base interface `Engine` or `Recipe` now. Give me a moment to think if we can potentially need that in the future.

```js
/**
 * Manual registration of the extension. Once calling `use` the auto discovery of extensions is turned off if not explicitly
 * turned on.
 * jsreport.use(require('jsreport-jsrender')())
 * @param {Object || Function} extensions
 * @return {Reporter} for chaining
 * @public
 */
Reporter.prototype.use = function (extension) {
  this.extensionsManager.use(extension)
  return this
}
```

File: types/jsreport-phantom-pdf/index.d.ts (hunk 4 above)
Comment (taoqf) [reply types/jsreport-phantom-pdf/index.d.ts]: OK, i will make another pr later. actually, I don't know much about this project, I declared the types just by searching the source code.

File: types/jsreport-phantom-pdf/index.d.ts (hunk 3 above)
Comment (pofider) [reply types/jsreport-phantom-pdf/index.d.ts]: `JSON.parse(true)` returns true. This code is there to assure both `"true"` and `true` are resolved as boolean. The reason for this is that the html form submits from studio are sending urlencoded values which are strings, not booleans. We make sure both are resolved as correct boolean.  Maybe jsreport-core types doesn't need to respect this and strictly demand boolean. Or at least `'true' | 'false' | booolean`.

File: types/jsreport-phantom-pdf/index.d.ts (hunk 4 above)
Comment (pofider) [reply types/jsreport-phantom-pdf/index.d.ts]: Nothing to worry about. I am not an expert in typescript. So great that we can finish this together. :) Thank you for your time.

File: types/jsreport-core/index.d.ts (hunk 1 above)
Comment (taoqf) [reply types/jsreport-core/index.d.ts]: Sorry, I didn't know that.

File: types/jsreport-html-to-xlsx/index.d.ts (hunk 2 above)
Comment (taoqf) [reply types/jsreport-html-to-xlsx/index.d.ts]: OK

File: types/jsreport-phantom-pdf/index.d.ts (hunk 3 above)
Comment (taoqf) [reply types/jsreport-phantom-pdf/index.d.ts]: OK

Comment (taoqf) [reply types/jsreport-phantom-pdf/index.d.ts]: Do you know how to define this?
```ts
blockJavaScript: 'true' | ???;
```
I could not figure out from the source.

File: types/jsreport-phantom-pdf/index.d.ts (hunk 4 above)
Comment (taoqf) [reply types/jsreport-phantom-pdf/index.d.ts]: It's all right. this project is so great that I just want to help more people.

File: types/jsreport-phantom-pdf/index.d.ts (hunk 3 above)
Comment (pofider) [reply types/jsreport-phantom-pdf/index.d.ts]: The `blockJavaScript: 'true' | 'false' | boolean` should work.
However, with deeper thinking. Please keep there just a boolean. We may remove support for passing strings in future. And it is better that typescript users are already instructed to send boolean.

File: types/jsreport-core/index.d.ts (hunk 5 above)
Comment (taoqf) [reply types/jsreport-core/index.d.ts]: I think if other people want a engine([dot](http://olado.github.io/doT/) is a good template engine as I know), they may want to know how to. and this interface Engine(if it has a exact type defination) may help them.

File: types/jsreport-phantom-pdf/index.d.ts (hunk 3 above)
Comment (taoqf) [reply types/jsreport-phantom-pdf/index.d.ts]: OK

File: types/jsreport-core/index.d.ts (hunk 5 above)
Comment (pofider) [reply types/jsreport-core/index.d.ts]: Registering engines and recipes is a little bit more complicated than your types defines.
The `use` function registers an extension, which is later invoked and can register an engine.
The correct signature of the `use` function is currently as this.

```js
type Extension = (reporter: Reporter, definition: object) => void;

interface ExtensionDefinition {
  options: any;
  main: any;
  directory: string;
}

interface Reporter {
  use(extension: Extension | ExtensionDefinition): Reporter;
}
```

The engine is then registered using
```js
jsreport.use((reporter : JsReport.Reporter) => {
  reporter.extensionsManager.engines.push({
    name: 'handlebars',
    pathToEngine: path.join(__dirname, 'handlebarsEngine.js')
  })
})
```
This is not so convenient at this moment, unfortunately.

Comment (taoqf) [reply types/jsreport-core/index.d.ts]: I see.

**Comment by pofider** [comment]
I think this version could be final for now. What you think?
I would like to take the changes here and send prs to DefinetelyTyped.
Create a dedicated folder there for each extension. As this:

`DefinitelyTyped/DefinitelyTyped/types/jsreport-core`
`DefinitelyTyped/DefinitelyTyped/types/jsreport-jsrender`
`DefinitelyTyped/DefinitelyTyped/types/jsreport-phantom-pdf`
`DefinitelyTyped/DefinitelyTyped/types/jsreport-html-to-xlsx`
`DefinitelyTyped/DefinitelyTyped/types/jsreport-xlsx`

Would you be able to do this?

**Comment by taoqf** [comment]
Now we have choice:

1. put all these types into one single file and maintain them in this project, at the same time, other people could add extends by themselves, like we did before(maybe we need a guide in readme).
2. put all these types into one single file and maintain them in DefinitelyTyped
3. put these types into separate files and maintain them in their own projects.
4. put these types into separate files and maintain them in DefinitelyTyped

I prefer 1, because I think it may stop contributors who want write a type definition or just send an issue when they have problems with the type definition in DefinitelyTyped, it's very strict in DefinitelyTyped.

I think DefinitelyTyped's purpose is to help people who want use a project in npm (or not in) without type definition in typescript, so if a project start support typescript in their own project, it should be removed in DefinitelyTyped.

Anyway ,choose one, I will do the work with pleasure.

**Comment by taoqf** [comment]
@pofider Have you seen this?

**Comment by taoqf** [comment]
I will be away for couple of days, I have commit a pr in DefinitelyTyped https://github.com/DefinitelyTyped/DefinitelyTyped/pull/23523

**Comment by pofider** [comment]
Thank you very much. I apologize for the delay. 

I did check many the top ranked packages in npm and the majority of them hosts the types in DefinitelyTyped. I understand it can be more difficult to maintain it there, but I would like to try it. If it will be pain for us, we can always remove it there and start to host types directly in our projects.

> put these types into separate files and maintain them in DefinitelyTyped

This approach with splitting types into folders for each particular extension sounds better. Because developers then get only types they need as they import extensions. The developer using 2 extensions gets exactly types he needs. Not all of them.

**Comment by taoqf** [comment]
I am back!!!
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/23848

**Comment by pofider** [comment]
Excellent! Thank you. The DT PR looks good.

**Comment by taoqf** [comment]
sadly, we could not export const enum.

**Comment by pofider** [comment]
Hm, I was not careful with checking the changes in DT. 
Why we could not export const enums? 
This is was the great part of it, so developer can dot and list recipes and engines.

**Comment by taoqf** [comment]
Because these packages do not export that enume actually, I mean in js, and exporting enum is not the suggested way in DT.
